### PR TITLE
`BlockId` removal: `CallApiAt::state_at`

### PR DESCRIPTION
**Pull request by michalkucharczyk**
It changes the arguments of:
- `CallApiAt::state_at`,
- `Client::code_at`,
- `Client::state_at`

from: `BlockId<Block>` to: `Block::Hash`

This PR is part of BlockId::Number refactoring analysis (https://github.com/paritytech/substrate/issues/11292)

### CODE REVIEW

File: client/service/src/client/client.rs [hunk 1]
```diff
@@ -1746,9 +1745,8 @@ where
 		CallExecutor::runtime_version(&self.executor, hash).map_err(Into::into)
 	}
 
-	fn state_at(&self, at: &BlockId<Block>) -> Result<Self::StateBackend, sp_api::ApiError> {
-		let hash = self.backend.blockchain().expect_block_hash_from_id(at)?;
-		self.state_at(hash).map_err(Into::into)
+	fn state_at(&self, at: <Block as BlockT>::Hash) -> Result<Self::StateBackend, sp_api::ApiError> {
```
Comment (bkchr) [inline client/service/src/client/client.rs:1748]:
```suggestion
	fn state_at(&self, at: Block::Hash) -> Result<Self::StateBackend, sp_api::ApiError> {
```

File: primitives/api/src/lib.rs [hunk 2]
```diff
@@ -622,7 +622,7 @@ pub trait CallApiAt<Block: BlockT> {
 	fn runtime_version_at(&self, at: &BlockId<Block>) -> Result<RuntimeVersion, ApiError>;
 
 	/// Get the state `at` the given block.
-	fn state_at(&self, at: &BlockId<Block>) -> Result<Self::StateBackend, ApiError>;
+	fn state_at(&self, at: <Block as BlockT>::Hash) -> Result<Self::StateBackend, ApiError>;
```
Comment (bkchr) [inline primitives/api/src/lib.rs:625]:
```suggestion
	fn state_at(&self, at: Block::Hash) -> Result<Self::StateBackend, ApiError>;
```

**Review by bkchr — APPROVED** [review]

**Comment by michalkucharczyk** [comment]
bot fmt

**Comment by michalkucharczyk** [comment]
bot fmt

**Comment by michalkucharczyk** [comment]
bot cancel

File: primitives/api/test/tests/runtime_calls.rs [hunk 3]
```diff
@@ -167,6 +166,7 @@ fn record_proof_works() {
 	}
 	.into_signed_tx();
 
+	let block_id = BlockId::Hash(client.chain_info().best_hash);
```
Comment (melekes) [inline primitives/api/test/tests/runtime_calls.rs:169]:
```suggestion
	let best_hash = BlockId::Hash(client.chain_info().best_hash);
```

**Review by melekes — APPROVED** [review]
👍

File: primitives/api/test/tests/runtime_calls.rs (hunk 3 above)
Comment (michalkucharczyk) [reply primitives/api/test/tests/runtime_calls.rs:169]: Typically `block_id` was always called `block_id`. Please also note that its type is not a `Hash`, it is a `BlockId`.

**Comment by michalkucharczyk** [comment]
bot rebase

**Comment by michalkucharczyk** [comment]
bot merge

**Pull request merged**